### PR TITLE
Heap allocate string literals

### PR DESCRIPTION
**Pull request by mppf**
For issues #18287 and #18284

This PR adjusts the compiler code that emits string literal initialization in two ways.

First, it adjusts the initialization for string literals to be emitted after dead string literals are removed. That simplifies the logic by removing the need to remove string literal initialization code in the event that the literal is unused. So, after this PR, until deadCodeElimination, string literals only exist as DefExprs for module-level variables in the ChapelStringLiterals module.

Second, it changes the initialization pattern for string literals to create a single heap allocation to store all string and bytes literals and then to copy the string literals from the read-only data section into this heap allocation. This avoids problems with communicating from the read-only data section for string literals.

For now, the single heap allocation storing the string and bytes literals is leaked. This leak is not reported by `--memLeaks`.

Testing:
- [x] multilocale tests pass on an Infiniband system with `CHPL_COMM=gasnet` and `CHPL_COMM_SEGMENT=everything` 
- [x] full local testing
- [x] full local gasnet testing

Reviewed by @e-kayrakli - thanks!

Future Work:
 - add serialize/deserialize for `bytes` and have the compiler check that module-scope `const`s are either POD or have serialize/deserialize implemented (see https://github.com/Cray/chapel-private/issues/2472 )
 - free the heap allocation storing the string literals in the module deinit
 - figure out how the corresponding functionality can work with separate compilation and/or dynamic loading of Chapel code

### CODE REVIEW

File: compiler/AST/symbol.cpp [hunk 1]
```diff
@@ -1508,6 +1511,111 @@ void createInitStringLiterals() {
...
+          Immediate* imm = s->immediate;
+          INT_ASSERT(imm);
+          literals.push_back({imm->to_string(), s});
+        }
+      }
+    }
+  }
+
+  // sort the strings/bytes so that they can be stored in a reliable order
+  std::sort(literals.begin(), literals.end());
```
Comment (e-kayrakli) [inline compiler/AST/symbol.cpp:1538]: Is this order still reliable if you have a Chapel string and a bytes that are identical? I think they'll end up being represented by different immediates, but they'll compare equal. So the ordering will rely on the VarSymbol pointers, which is indeterministic.

File: modules/internal/String.chpl [hunk 2]
```diff
@@ -602,11 +602,27 @@ module String {
   }
 
   pragma "no doc"
-  proc chpl_createStringWithLiteral(x: c_string, length: int, numCodepoints: int) {
+  proc chpl_createLiteralsBuffer(size: int): c_void_ptr {
+    return c_malloc(uint(8), size):c_void_ptr;
```
Comment (e-kayrakli) [inline modules/internal/String.chpl:606]: What's the challenge for freeing this? I would imagine putting this allocation in a module level variable and freeing that in module deinitializer should work, but do we end up deallocating the buffer too early if we do that?

File: compiler/AST/symbol.cpp (hunk 1 above)
Comment (mppf) [reply compiler/AST/symbol.cpp:1538]: That's mostly true, but we have defined `std::less` for the VarSymbol pointers. See

https://github.com/chapel-lang/chapel/blob/c75946acea8a1f98acd6235cb09e72046baf8254/compiler/include/baseAST.h#L438-L448 (which is later invoked as `def_less_ast(VarSymbol)`

We have quite a few `std::map`s that rely on this to have reproducible ordering.

File: modules/internal/String.chpl (hunk 2 above)
Comment (mppf) [reply modules/internal/String.chpl:606]: I would also expect it to work but I'd like to try that in a different PR at the very least. I'm open to feedback for whether or not we should try to do it for this release. Anyway I am worried that we might run in to an issue but it should be easy enough to try what you are describing.

File: compiler/AST/symbol.cpp (hunk 1 above)
Comment (e-kayrakli) [reply compiler/AST/symbol.cpp:1538]: Ah, thanks!

File: modules/internal/String.chpl (hunk 2 above)
Comment (e-kayrakli) [reply modules/internal/String.chpl:606]: As a generally cautious person around the release dates, I feel brave about trying to do this for the release. A full ASAN test with local and gasnet configs would make me feel more comfortable about it. And if that testing turns up some bugs, we can punt it.

That being said, this leak is not too worrisome to me, as it is not something that would grow dynamically or anything. It'll be an O(n) leak where n is the amount of string literals used in the code. And we'll be freeing it at the teardown of the application, so it shouldn't impact the "steady state" memory footprint of the application in any way.

So, I say we go for it if you have the time. But I'm happy with whatever you decide at the end.

Comment (mppf) [reply modules/internal/String.chpl:606]: I've created #18319 for this.

**Pull request merged**